modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/segstore/GroupIndexMeta.java

```diff
@@ -19,57 +19,117 @@
-        IndexFileMetaArray fileMetas = this.fileMetas;
+        IndexMetaArrayHolder curFileMetas = fileMetaDeque.getLast();
+
+        long curLastLogIndex = curFileMetas.fileMetas.lastLogIndexExclusive();
 
-        IndexFileMetaArray newFileMetas = fileMetas.add(indexFileMeta);
+        long newFirstLogIndex = indexFileMeta.firstLogIndexInclusive();
 
-        // Simple assignment would suffice, since we only have one thread writing to this field, but we use compareAndSet to verify
-        // this invariant, just in case.
-        boolean updated = FILE_METAS_VH.compareAndSet(this, fileMetas, newFileMetas);
+        assert newFirstLogIndex <= curLastLogIndex :
+                String.format(
+                        "Gaps between Index File Metas are not allowed. Last log index: %d, new log index: %d",
+                        curLastLogIndex, newFirstLogIndex
+                );
 
-        assert updated : "Concurrent writes detected";
+        // Merge consecutive index metas into a single meta block. If there's an overlap (e.g. due to log truncation), start a new block,
+        // which will override the previous one during search.
+        if (curLastLogIndex == newFirstLogIndex) {
+            curFileMetas.addIndexMeta(indexFileMeta);
+        } else {
+            fileMetaDeque.add(new IndexMetaArrayHolder(indexFileMeta));
+        }
     }
 
     /**
-     * Returns a file pointer that uniquely identifies the index file for the given log index. Returns {@code null} if the given log index
+     * Returns index file meta that uniquely identifies the index file for the given log index. Returns {@code null} if the given log index
      * is not found in any of the index files in this group.
      */
     @Nullable
     IndexFileMeta indexMeta(long logIndex) {
-        return fileMetas.find(logIndex);
+        Iterator<IndexMetaArrayHolder> it = fileMetaDeque.descendingIterator();
+
+        while (it.hasNext()) {
+            IndexFileMetaArray fileMetas = it.next().fileMetas;
+
+            // Log suffix might have been truncated, so we can have an entry on the top of the queue that cuts off part of the search range.
+            if (logIndex >= fileMetas.lastLogIndexExclusive()) {
+                return null;
+            }
+
+            if (logIndex < fileMetas.firstLogIndexInclusive()) {
+                continue;
+            }
+
+            IndexFileMeta indexMeta = fileMetas.find(logIndex);
+
+            if (indexMeta != null) {
+                return indexMeta;
+            }
+        }
+
+        return null;
     }
 
     long firstLogIndexInclusive() {
-        return fileMetas.get(0).firstLogIndexInclusive();
+        for (IndexMetaArrayHolder indexMetaArrayHolder : fileMetaDeque) {
+            IndexFileMetaArray fileMetas = indexMetaArrayHolder.fileMetas;
+
+            if (fileMetas.size() > 0) {
+                return fileMetas.firstLogIndexInclusive();
+            }
```

Comment:
I believe that this condition is only needed to differentiate between an empty log and a non-empty log. Is that true? If the answer is yes then could you please add a small clarifying comment to the code? If the answer is no then maybe it requires a larger comment. Thank you!

EDIT: Now that I think about it, I don't really understand what exactly does the `size` method return. Size of the array? How can it be zero?